recognition/include/pcl/recognition/impl/linemod/line_rgbd.hpp

```diff
@@ -150,72 +151,68 @@ LineRGBD<PointXYZT, PointRGBT>::loadTemplates (const std::string &file_name, con
-    PointCloud<PointXYZRGBA> & template_point_cloud = template_point_clouds_[i];
-    BoundingBoxXYZ & bb = bounding_boxes_[i];
-    bb.x = bb.y = bb.z = std::numeric_limits<float>::max ();
-    bb.width = bb.height = bb.depth = 0.0f;
-
-    float center_x = 0.0f;
-    float center_y = 0.0f;
-    float center_z = 0.0f;
-    float min_x = std::numeric_limits<float>::max ();
-    float min_y = std::numeric_limits<float>::max ();
-    float min_z = std::numeric_limits<float>::max ();
-    float max_x = -std::numeric_limits<float>::max ();
-    float max_y = -std::numeric_limits<float>::max ();
-    float max_z = -std::numeric_limits<float>::max ();
-    std::size_t counter = 0;
-    for (std::size_t j = 0; j < template_point_cloud.size (); ++j)
-    {
-      const PointXYZRGBA & p = template_point_cloud.points[j];
+    computeBoundingBox (i);
+  }
 
-      if (!std::isfinite (p.x) || !std::isfinite (p.y) || !std::isfinite (p.z))
-        continue;
+  return (true);
+}
 
-      min_x = std::min (min_x, p.x);
-      min_y = std::min (min_y, p.y);
-      min_z = std::min (min_z, p.z);
-      max_x = std::max (max_x, p.x);
-      max_y = std::max (max_y, p.y);
-      max_z = std::max (max_z, p.z);
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+template <typename PointXYZT, typename PointRGBT> void
+pcl::LineRGBD<PointXYZT, PointRGBT>::computeBoundingBox (int i)
+{
+  PointCloud<PointXYZRGBA> & template_point_cloud = template_point_clouds_[i];
+  BoundingBoxXYZ & bb = bounding_boxes_[i];
+  bb.x = bb.y = bb.z = std::numeric_limits<float>::max ();
+  bb.width = bb.height = bb.depth = 0.0f;
+
+  Eigen::Vector3f geometric_center = Eigen::Vector3f::Zero ();
+  Eigen::Vector3f min_pos, max_pos;
+  min_pos.fill (std::numeric_limits<float>::max ());
+  max_pos.fill (std::numeric_limits<float>::lowest ());
+  std::size_t counter = 0;
+  for (std::size_t j = 0; j < template_point_cloud.size (); ++j) 
+  {
+    const PointXYZRGBA & p = template_point_cloud.points[j];
 
-      center_x += p.x;
-      center_y += p.y;
-      center_z += p.z;
+    if (!isXYZFinite(p))
+      continue;
 
-      ++counter;
+    for (std::size_t row=0; row<3; ++row)
+    {
+      min_pos[row] = std::min (min_pos[row], p.getVector3fMap ()[row]);
+      max_pos[row] = std::max (max_pos[row], p.getVector3fMap ()[row]);
     }
 
-    center_x /= static_cast<float> (counter);
-    center_y /= static_cast<float> (counter);
-    center_z /= static_cast<float> (counter);
+    geometric_center += p.getVector3fMap ();
 
-    bb.width  = max_x - min_x;
-    bb.height = max_y - min_y;
-    bb.depth  = max_z - min_z;
+    ++counter;
+  }
+  geometric_center /= static_cast<float> (counter);
 
-    bb.x = (min_x + bb.width / 2.0f) - center_x - bb.width / 2.0f;
-    bb.y = (min_y + bb.height / 2.0f) - center_y - bb.height / 2.0f;
-    bb.z = (min_z + bb.depth / 2.0f) - center_z - bb.depth / 2.0f;
+  auto bb_dim = max_pos - min_pos;
```

Comment:
No auto without eval since Eigen uses lazy return type.

Reply:
I'm not sure what you mean by lazy return type. Could you explain that to me please?

Reply:
@kunaltyagi

Reply:
See https://eigen.tuxfamily.org/dox/TopicLazyEvaluation.html for the explanation of Eigen's lazy evaluation, and https://eigen.tuxfamily.org/dox/TopicPitfalls.html for the pitfall of using `auto` with it.